loopy/translation_unit.py

```diff
@@ -480,20 +444,27 @@ def map_kernel(self, kernel):
 
         return callables_in_insn
 
+    def map_type_cast(self, expr):
+        return self.rec(expr.child)
+
     map_variable = map_constant
     map_function_symbol = map_constant
     map_tagged_variable = map_constant
-    map_type_cast = map_constant
 
 
 def _get_callable_ids_for_knl(knl, callables):
     clbl_id_collector = CallablesIDCollector()
 
-    return frozenset().union(*(
-        _get_callable_ids_for_knl(callables[clbl].subkernel, callables) |
-        frozenset([clbl]) if isinstance(callables[clbl], CallableKernel) else
-        frozenset([clbl])
-        for clbl in clbl_id_collector.map_kernel(knl)))
+    def rec(clbl_id):
+        clbl = callables[clbl_id]
+        if isinstance(clbl, CallableKernel):
+            return (_get_callable_ids_for_knl(clbl.subkernel, callables)
+                    | frozenset([clbl_id]))
+        else:
+            return frozenset([clbl_id])
```

Comment:
I feel like `get_called_callables` (or some such) should be a method on the `InKernelCallable`, to avoid this `isinstance` hack here. After all, callables other than `CallableKernel` could call further callables.

Reply:
e8bf9f1972f4733c8b3238cad26ccf9695e79232